modules/SantoriniBoard.class.php

```diff
@@ -57,45 +61,70 @@ public static function addInfo($piece)
     } elseif (substr($piece['type'], 0, 5) == 'token') {
       $piece['direction'] = $piece['type_arg'];
     }
+    Utils::convertIntValues($piece);
     return $piece;
   }
 
   /*
    * getPiece: return all info about a piece
-   * params : int $id
+   * params : mixed $id, either int or array with ['id'] key
+   */
+  public function getPiece($mixed)
+  {
+    $id = null;
+    if (is_numeric($mixed)) {
+      $id = $mixed;
+    } else if (is_array($mixed) && array_key_exists('id', $mixed)) {
+      $id = $mixed['id'];
+    }
+    if ($id == null) {
+      return null;
+    }
+    return self::addInfo(self::getObjectFromDB("SELECT * FROM piece WHERE id = '$id'"));
+  }
+
+  /*
+   * getPiecesAt: return array the pieces at this x,y,z location.
+   * tokens make it possible to have multiple pieces (e.g., Clio coin + worker)
+   * params : array $space
    */
-  public function getPiece($id)
+  public function getPiecesAt($space)
   {
-    return self::addInfo(self::getNonEmptyObjectFromDB("SELECT * FROM piece WHERE id = '$id'"));
+    return array_map('SantoriniBoard::addInfo', self::getObjectListFromDb("SELECT * FROM piece WHERE location = 'board' AND x = {$space['x']} AND y = {$space['y']} AND z = {$space['z']} ORDER BY id"));
   }
 
   /*
-   * getPieceAt: return all info about a piece at a location
+   * getBlocksAt: return array of blocks at this x,y location (for all z, order top down)
    * params : array $space
    */
-  public function getPieceAt($space)
+  public function getBlocksAt($space)
   {
-    return self::addInfo(self::getObjectFromDB("SELECT * FROM piece WHERE location = 'board' AND x = {$space['x']} AND y = {$space['y']} AND z = {$space['z']}"));
+    return array_map('SantoriniBoard::addInfo', self::getObjectListFromDb("SELECT * FROM piece WHERE location = 'board' AND type IN ('lvl0', 'lvl1', 'lvl2') AND x = {$space['x']} AND y = {$space['y']} ORDER BY z DESC"));
   }
 
   /*
    * getPiecesByType: return all info about pieces of the given type
    */
-  public function getPiecesByType($type, $type_arg = null)
+  public function getPiecesByType($type, $type_arg = null, $location = null)
   {
     $sql = "SELECT * FROM piece WHERE type = '$type'";
     if ($type_arg) {
       $sql .= " AND type_arg = '$type_arg'";
     }
+    if ($location) {
+      $sql .= " AND location = '$location'";
+    }
+    $sql .= " ORDER BY id";
     return array_map('SantoriniBoard::addInfo', self::getObjectListFromDb($sql));
   }
 
   /*
    * getPlacedPieces: return all pieces on the board
+   * Order by tokens first, then by z-order ascending
```

Comment:
The order is important now, because board.js looks at existing pieces in the same x,y,z when you add a new piece, to check if height needs to be adjusted for a token. We must put the token into the space before we put the worker into the space.